fix!: Don't gossip txs with duplicate public inputs

sirasistant opened the pull request:
Creates a new type for a CIVC proof with public inputs so cases like this are clear in the future. The Tx now uses the type without public inputs, and the public inputs are attached to it as needed.

sirasistant commented on yarn-project/stdlib/src/proofs/client_ivc_proof.ts:
```diff
@@ -46,12 +55,56 @@ export class ClientIvcProof {
+  constructor(
+    // The proof fields with public inputs.
+    // For recursive verification, use without public inputs via `removePublicInputs()`.
+    public fieldsWithPublicInputs: Fr[],
```
not calling this `fields` so TS's duck typing complains if we use one in place of the other.

LeilaWang commented on yarn-project/stdlib/src/proofs/client_ivc_proof.ts:
```diff
@@ -46,12 +55,56 @@ export class ClientIvcProof {
+  }
+
+  toJSON() {
+    return this.toBuffer();
```
This is wrong and probably not used anywhere!

sirasistant replied:
true! I just copied it over when duplicating the class. deleting!

LeilaWang reviewed (approved):
Beautiful! 🚀

spalladino commented on yarn-project/stdlib/src/kernel/private_kernel_tail_circuit_public_inputs.ts:
```diff
@@ -184,6 +184,12 @@ export class PrivateKernelTailCircuitPublicInputs {
     );
   }
 
+  publicInputs(): PrivateToPublicKernelCircuitPublicInputs | PrivateToRollupKernelCircuitPublicInputs {
+    return this.forPublic
+      ? this.toPrivateToPublicKernelCircuitPublicInputs()
+      : this.toPrivateToRollupKernelCircuitPublicInputs();
+  }
```
What's the cheapest way to add a unit test that throws if these public inputs no longer match the ones expected by the ClientIVC proof?

sirasistant replied:
Depends on what you mean with match:
 - If you mean that given a real clientivc proof and some public inputs, that they match, that's not possible to do in a unit test, that's checked in the full prover e2e test, because it needs many systems (kernels, bb, etc)
 - If you mean that the shape matches, we can probably make a unit test for that pretty easily. I can see if there is something already in place and add it if it's not

spalladino replied:
Both. The full prover e2e should be the catch-all, but for a quick shape check, it'd be useful to have a unit test. Especially if we miss a case in the prover test, or it is unexpectedly silenced :-/

sirasistant replied:
Hmmmm yep a quick unit test to check that the serialization to fields on both sides (noir and ts) matches could be useful... it would need some nonempty fixtures to maintain synced on both sides though...

sirasistant replied:
I've been experimenting with this and I think we'd need a factories.nr, a version of factories.ts to maintain in sync. Either that or to autogenerate huge fixtures from the TS factories. Then we can compare the serialization of the same object. It's unclear to me if it's worth the friction...

spalladino commented on yarn-project/pxe/src/private_kernel/private_kernel_execution_prover.ts:
```diff
@@ -352,7 +352,7 @@ export class PrivateKernelExecutionProver {
     if (!skipProofGeneration) {
       const provingTimer = new Timer();
-      clientIvcProof = await this.proofCreator.createClientIvcProof(executionSteps);
+      clientIvcProof = (await this.proofCreator.createClientIvcProof(executionSteps)).removePublicInputs();
```
Should we add a check here, before sending the tx, that the public inputs we're removing match the ones from the `Tx`?

sirasistant replied:
Actually, that's why I didn't remove the public inputs inside the proof creator, just in case we wanted to do that. Will see if it's simple enough to add in this PR!

sirasistant replied:
(added this)

spalladino reviewed (approved):
Thanks for tackling this!

sirasistant committed:
fix!: Don't gossip txs with duplicate public inputs

Creates a new type for a CIVC proof with public inputs so cases like this are clear in the future. The Tx now uses the type without public inputs, and the public inputs are attached to it as needed.

sirasistant commented:
ugh, one of the machines was evicted and had to start over, and timed out http://ci.aztec-labs.com/7aa0534eb0e3c304
